src/Features/CSharp/Portable/CodeFixes/GenerateMethod/GenerateDeconstructMethodCodeFixProvider.cs

```diff
@@ -0,0 +1,89 @@
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.GenerateMember.GenerateParameterizedMember;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.GenerateDeconstructMethod
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.GenerateDeconstructMethod), Shared]
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.GenerateEnumMember)]
+#pragma warning disable RS1016 // Code fix providers should provide FixAll support. https://github.com/dotnet/roslyn/issues/23528
+    internal class GenerateDeconstructMethodCodeFixProvider : CodeFixProvider
+#pragma warning restore RS1016 // Code fix providers should provide FixAll support.
+    {
+        private const string CS8129 = nameof(CS8129); // No suitable Deconstruct instance or extension method was found...
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(CS8129);
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            // Not supported in REPL
+            if (context.Project.IsSubmission)
+            {
+                return;
+            }
+
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var span = context.Span;
+            var token = root.FindToken(span.Start);
+
+            var deconstruction = token.GetAncestors<SyntaxNode>()
+                .FirstOrDefault(n => n.IsKind(SyntaxKind.SimpleAssignmentExpression, SyntaxKind.ForEachVariableStatement));
+
+            if (deconstruction is null)
+            {
+                Debug.Assert(false);
+                return;
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            DeconstructionInfo info;
+            ITypeSymbol type;
+            ExpressionSyntax target;
+            switch (deconstruction)
+            {
+                case ForEachVariableStatementSyntax @foreach:
+                    info = model.GetDeconstructionInfo(@foreach);
+                    type = model.GetForEachStatementInfo(@foreach).ElementType;
+                    target = @foreach.Variable;
+                    break;
+                case AssignmentExpressionSyntax assignment:
+                    info = model.GetDeconstructionInfo(assignment);
+                    type = model.GetTypeInfo(assignment.Right).Type;
+                    target = assignment.Left;
+                    break;
+                default:
+                    throw ExceptionUtilities.Unreachable;
+            }
```

Comment:
nit, consider a helper local or silbing function.  That way you could actually use a deconstruction here and just do:

```var (info, type, target) = Whatever()```, and in that Whatever method, you could more eaisly just return those values.  Just an idea :)

Reply:
I'll leave as-is. The method is pretty simple overall.

Reply:
Works for me.

Reply:
@alrz once we have match-expressions, it would be great to have a refactoring that finds these and offers to convert them :)